### DISCUSSION TITLE
Experimental: dark mode / themes

### DISCUSSION THREAD

**Discussion by kjk** [sumatrapdfreader/sumatrapdf#3432] [Announcements]
Thanks to work of @avorobey we now have experimental support for dark theme.

You can get it in latest pre-release version: https://www.sumatrapdfreader.org/prerelease

To use it:
* `Ctrl + K` to open Command Palette
* type `next theme` to invoke `Switch to next theme command`

To report issues related to dark mode: https://github.com/sumatrapdfreader/sumatrapdf/issues/3429

**Comment by omanuelcosta** [comment]
Amazing! Finally!

How can I make it permanently, every time I restart it, it goes back to white theme.

**Comment by selden** [comment]
@omanuelcosta 
One way to make "dark mode" permanent is to open the menu "Settings / Advanced Options..." and changing 
TextColor from 000000 to eeeeee
and 
BackgroundColor from ffffff to 333333

Other color values might work as well, but these are the values that I like.

**Reply by GitHubRulesOK** [reply]
thats not theming your inverting colours and can do that from command line for permanent canvas changes or simpler press [I]nvert button

**Comment by selden** [comment]
Bug in "dark mode":

When using "dark mode" to view epub documents, the colors of all included images are inverted. Dark colors become light and light colors become dark. They should not be. Only text and background colors should change. Images should not be affected.

This problem is seen in v 3.5.15481 (prerelase) when  using Ctrl-K / Switch to next theme
and in both v 3.5.15481 and v 3.4.6
when changing the values of TextColor and BackgoundColor.

The screengrabs below were taken under Windows 10 using "Switch to next theme" with  SumatraPDF v3.5.15481

Light mode, showing correct image coloration:

![light_mode](https://github.com/sumatrapdfreader/sumatrapdf/assets/353932/cc184fcb-3fe4-4684-a9ed-837fe8416601)

Dark mode showing incorrect image coloration:

![dark_mode](https://github.com/sumatrapdfreader/sumatrapdf/assets/353932/18d007ce-82ba-439b-a766-45192124446a)

Zip containing minimal epub demonstrating the bug:
[MyStory.zip](https://github.com/sumatrapdfreader/sumatrapdf/files/12362603/MyStory.zip)

**Reply by GitHubRulesOK** [reply]
@selden  Inversion is exactly that, turn all pink faces into smurf blue or turn white paper into black so there is no problem or issue with MuPDF inversion it is the same in any paint package invert means make a negative colour image. There was a different method used for legacy (older ) ebook reader where images were kept seperate loading from the page contents but blue text should have turned orange and brown images blue. the bug was in UI could not show page numbers and other white text so that is "fixed" by UI theming but inversion is a historic non theming issue that to turn all colours upside down pleases no one.
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/5669b36d-4067-4592-a463-e168a24d90ce)

**Comment by selden** [comment]
(Edited to add pdf document.)

While inverting colors is an easy way to approximate "dark mode",  "dark mode" and "invert colors" are NOT the same. There are other document readers  which implement "dark mode" as I've described: they reverse text coloration but preserve image coloration. It'd be really nice if SumatraPDF did the same.

An example is the free edition of "PDF-XChange Editor" by Tracker Software Products. In dark mode, it displays PDF documents with a dark backgound and white text but preserves the colors of embedded images. Unfortunately, it can't display epub documents.

Another slightly different example is how "dark mode" is implemented by these GitHub pages when viewed in Firefox or Chrome. Again, text is shown as white letters on a black background, while image colors are unchanged.

As an example, below are screengrabs of a document which I created in LibreOffice Writer and exported to PDF. It contains text and an image. (I reduced the sizes of the screengrabs.)

The PDF document:
[document_with_image.pdf](https://github.com/sumatrapdfreader/sumatrapdf/files/12364269/document_with_image.pdf)

As seen while editing in Writer (with dark mode UI)
![writer](https://github.com/sumatrapdfreader/sumatrapdf/assets/353932/dbbafd59-dd84-4b23-926f-464600354bda)


As shown by PDF-XChange Editor in "dark mode"
![PDF-Xchange](https://github.com/sumatrapdfreader/sumatrapdf/assets/353932/8a847f53-576b-4873-bec2-ee6e61800f45)


As shown by SumatraPDF in "next theme"
![SumatraPDF](https://github.com/sumatrapdfreader/sumatrapdf/assets/353932/359d484c-50e5-4827-8469-1869a17cf236)

**Comment by xbowzone** [comment]
Excluding images from color inversion, can this be implemented?
Edit: But I also would like for image color inversion to remain as well as an option. I often encounter pdfs which are entire books scanned and saved as images and not text, and sumatra has been my go to in those situations but for the pdfs that're not like this(most of them), image inversion has a negative reading experience.

**Comment by technodrome** [comment]
I also vote for this feature - having images render in their original non-inverted state while the rest is dark. It's painful on the eyes to have LSD-like images with all kinds of crazy pink when reading at night. I strongly believe this would be a very nice selling point for Sumatra. Every other app can do simple eyesore-inducing full-scale inversion.

**Comment by selden** [comment]
For the moment I'm using Calibre to read ePubs in dark mode, but it comes with a *lot* of baggage. I'd really like to be able to use Sumatra: the ease of use of its user interface is far better.

**Comment by gsirvine** [comment]
Plus 1 on not inverting images 

Similar to how the Dark Reader extension does in web browsers. I know it's possible because before firefox made a version years ago that said you couldn't use extensions on pdf documents I simply used Dark Reader just like I did on websites. PDF documents looked amazing. I don't think the dark reader code was that complicated at the time, i think that was before they created the dynamic mode so it was just filter and filter+. 

You might even be able to reach out to the dark reader dev on a solution to this.

**Comment by Andy2No** [comment]
Is there a way to get the bookmarks sidebar in dark mode, at the same time as the document, in SumatraPDF-3.5.2?

I previously had that working fine in v3.4.6, which I've been using up until now.  Installing 3.5.2 over it has broken it.  If I set the theme to dark or darker, in 3.5.2, I get a dark bookmarks sidebar, but the document is then on a white background.  If I set it to Light, I get a dark background for the document but a white, eye straining, background for the bookmarks.

I'm not clear if there was a theme menu in v3.4.6 - possibly I just did it all through the Advanced Settings text file.  Is so, is 3.5.2 inverting my settings, to make the new dark mode?

I think I'll revert to v3.4.6, for now, so no rush for a reply.  FWIW, here's my SumatraPDF-settings.txt contents;

```
# For documentation, see https://www.sumatrapdfreader.org/settings/settings3-4-6.html
FixedPageUI [
	TextColor = #f0ff00
	BackgroundColor = #002000
	SelectionColor = #ff00ff
	WindowMargin = 0 0 0 0
	PageSpacing = 4 4
	HideScrollbars = false
]
ComicBookUI [
	WindowMargin = 0 0 0 0
	PageSpacing = 4 4
	CbxMangaMode = false
]
ChmUI [
	UseFixedPageUI = false
]

SelectionHandlers [
]
ExternalViewers [
]

ZoomLevels = 8.33 12.5 18 25 33.33 40 50 60 66.67 70 72.5 75 80 85 90 95 100 105 110 115 120 125 150 175 200 250 300 350 400 500 600 800 1000 1200 1600
ZoomIncrement = 0

PrinterDefaults [
	PrintScale = shrink
]
ForwardSearch [
	HighlightOffset = 0
	HighlightWidth = 15
	HighlightColor = #6581ff
	HighlightPermanent = false
]
Annotations [
	HighlightColor = #ffff00
	UnderlineColor = #00ff00
	TextIconColor = #ffff00
	TextIconType = 
	DefaultAuthor = 
]

RememberOpenedFiles = true
RememberStatePerDocument = true
RestoreSession = true
UiLanguage = en
EnableTeXEnhancements = false
DefaultDisplayMode = continuous
DefaultZoom = fit page
Shortcuts [
]
EscToExit = false
ReuseInstance = false
ReloadModifiedDocuments = true

MainWindowBackground = #80002200
FullPathInTitle = false
ShowMenubar = true
ShowToolbar = true
ShowFavorites = false
ShowToc = true
TocDy = 0
SidebarDx = 150
ToolbarSize = 18
TabWidth = 100
TreeFontSize = 0
SmoothScroll = false
ShowStartPage = true
CheckForUpdates = false
WindowState = 1
WindowPos = 360 0 1002 730
UseTabs = false
UseSysColors = false
CustomScreenDPI = 0

FileStates [

 [ REDACTED ]

]
SessionData [
]
TimeOfLastUpdateCheck = 0 0
OpenCountWeek = 681

# Settings below are not recognized by the current version
```

**Comment by username116** [comment]
If one day you want to make a theme or an option to "Follow system theme" automatically, maybe this registry key can serve :
`HKEY_CURRENT_USER\SOFTWARE\Microsoft\Windows\CurrentVersion\Themes\Personalize`
`AppsUseLightTheme` is 0 or 1 depending on the system theme (I'm on Windows 10). [source](https://www.tenforums.com/tutorials/24038-change-default-app-windows-mode-light-dark-theme-windows-10-a.html#option2)

Thank you very much for making it possible to set the colors in the dark theme (with https://github.com/sumatrapdfreader/sumatrapdf/commit/5ceb24aadd2021a6ee2e4e14a3589fa65202790b). This is something I was missing. 💛

**Comment by Programmer96** [comment]
Theme = Darker
To make it permanent

**Comment by Grovkillen** [comment]
I find the "active tab" color being not very clear when using the Dark (or Darker) theme.

**Reply by tamr** [reply]
For me it's completely unreadable. Not a big deal but sometimes requires few extra clicks to check the chapter's title.

![23910 copy](https://github.com/user-attachments/assets/04b02441-e56c-4d51-8bfc-e057aeb62a25)

**Comment by gsirvine** [comment]
any update on dark mode not inverting images?

**Reply by GitHubRulesOK** [reply]
Generally impossible  (for RGBA/CMYK Images and thus PDF) the word inversion means just that in black & white terms. 
Reverse the core colours means reverse all the colours there is NO DARK/NIGHT MODE except reduce HARD BLUE LIGHTS and use RED PAPER.
![image](https://github.com/user-attachments/assets/281e3419-aa68-4a81-923f-0033f5aab97f)

WHY is it a problem is the images are simply inline in a page and thus the colors are set for a page text or image reverse will be similar white text will become black text and flesh colour faces become smurf like blues.

To counteract the image inversion the images need to be **edited** into their opposing colours before loading into a reversed page 
![image](https://github.com/user-attachments/assets/20b4c423-6223-4e42-9148-e7cd810658bc)


One such editor is possibly (NOT endorsed but offered as an example)  https://pdfcandle.com/pdf_invert.aspx
